### ISSUE TITLE
TMRegion: also output predictiveCells

### ISSUE THREAD

**Issue by breznak** [htm-community/nupic.cpp#412]
we can output activeCells, predictedActiveCells (winnerCells), but not predictiveCells. 
Use tm.getPredictiveCells() to obtain the data.

**Comment by dkeeney** [comment]
So, what should be output for `"bottomUpOut"` in TMRegion?
The description only says `"The output signal generated from the bottom-up inputs from lower levels.   The width is 'numberOfCols' * 'cellsPerColumn'."`.

If NetworkAPI is to be 100% compatible with original implementations then its outputs should be the same.

**Comment by breznak** [comment]
> The description only says "The output signal generated from the bottom-up inputs from lower levels. The width is 'numberOfCols' * 'cellsPerColumn'.".

Isn't this description given by us? I mean TMRegion in C++ did not exist, till you've implemented it. Or does it match python side? 

> does it match python side?  

If so, what's for bottomUpOutput? 
If not, I'd say bottomUpOut could be predictive cells (as columns?). 
 - as active cells already match SPRegion's output. 
So the default (bottomUpOut) is one thing, the other is providing output name "predictiveCells".

**Comment by dkeeney** [comment]
TMRegion was a port of python code (or at least attempted to do so).  The text came from there although it might have been modified.  The point is it does not say what the output actually is.

**Comment by dkeeney** [comment]
The other point is that "BottomUpOut" should be whatever the original Python implementation provided since that is what the NetworkAPI user will be expecting.

**Comment by breznak** [comment]
> The other point is that "BottomUpOut" should be whatever the original Python implementation provided since that is what the NetworkAPI user will be expecting. 

ok, I looked at the `tm_region.py`, and it's a mess. It allows to use many (cpp/py, TM, BacktrackingTM) "temporal memory" implementations, and the region itself would break if other than BackTM is used. So that implementation is broken (more then ours). 

- [x] the API/syntax would not break, we still provide field "bottomUpOutput", so it's interchangeable. 
- [x] the API is documented/updated with what the field means (was not clear before anyway),
- [x] BackTM (and old TP) have configurable `outputMode` (normal = default, OR(active, predictive) = what we removed as flawed; activeCells, 1cellPerColumn, ...) one of these is conveied by orColumnOutputs/now better named asColumns. 
- [ ] the runtime output may be different to other implementations (because we fix what is broken, BackTM removed as flawed, etc) 

TL;DR: I think we must keep providing "bottomUpOutput", but are relatively free to define what its output means.

**Comment by dkeeney** [comment]
> ok, I looked at the tm_region.py, and it's a mess.

Thanks for taking a look at the py code and your assessment of its condition is the same as mine.  From py you can still get to our C++ implementation of TM but it means going through the BacktrackingTM py code, then through that shim logic to convert BacktrackingTM's parameters into TM's parameters.  It is a real mess.

Of course with NetworkAPI all of that can be bypassed now that we have TMRegion but we need to document that in the wiki.
